apps/emqx/src/emqx_limiter/src/emqx_limiter_shared/emqx_limiter_allocator.erl

```diff
@@ -0,0 +1,346 @@
+
+-spec update(emqx_limiter:group()) -> ok.
+update(Group) ->
+    gen_server:call(?VIA_GPROC(Group), #update{}, infinity).
+
+%%--------------------------------------------------------------------
+%%% gen_server callbacks
+%%--------------------------------------------------------------------
+
+-spec init(list()) -> {ok, state()}.
+init([Group, LimiterNames]) ->
+    process_flag(trap_exit, true),
+    State0 = #{
+        group => Group,
+        counter => counters:new(length(LimiterNames), [write_concurrency]),
+        buckets => #{},
+        last_alloc_time => now_ms_monotonic(),
+        timers => #{},
+        burst_timers => #{}
+    },
+    State1 = create_buckets(State0, LimiterNames),
+    State2 = ensure_alloc_timers(State1),
+    State3 = ensure_burst_timers(State2),
+    {ok, State3}.
+
+handle_call(#update{}, _From, State0) ->
+    State1 = ensure_alloc_timers(State0),
+    State2 = ensure_burst_timers(State1),
+    {reply, ok, State2};
+handle_call(Req, _From, State) ->
+    ?SLOG(error, #{msg => "emqx_limiter_allocator_unexpected_call", call => Req}),
+    {reply, ignored, State}.
+
+handle_cast(Req, State) ->
+    ?SLOG(error, #{msg => "emqx_limiter_allocator_unexpected_cast", cast => Req}),
+    {noreply, State}.
+
+handle_info(#tick_alloc_event{interval = Interval}, State) ->
+    {noreply, handle_alloc_timer(State, Interval)};
+handle_info(#tick_burst_alloc_event{interval = Interval}, State) ->
+    {noreply, handle_burst_timer(State, Interval)};
+handle_info(Info, State) ->
+    ?SLOG(error, #{msg => "emqx_limiter_allocator_unexpected_info", info => Info}),
+    {noreply, State}.
+
+terminate(_Reason, #{group := Group, buckets := Buckets} = _State) ->
+    maps:foreach(
+        fun(LimiterName, _) ->
+            LimiterId = {Group, LimiterName},
+            emqx_limiter_bucket_registry:delete_bucket(LimiterId)
+        end,
+        Buckets
+    ),
+    ok.
+
+%%--------------------------------------------------------------------
+%%% Internal functions
+%%--------------------------------------------------------------------
+
+create_buckets(State, LimiterNames) ->
+    {_, Buckets} = lists:foldl(
+        fun(LimiterName, {Index, BucketsAcc}) ->
+            Bucket = create_bucket(
+                State, LimiterName, Index
+            ),
+            {Index + 1, BucketsAcc#{LimiterName => Bucket}}
+        end,
+        {1, #{}},
+        LimiterNames
+    ),
+    State#{buckets => Buckets}.
+
+create_bucket(#{group := Group, counter := Counter}, LimiterName, Index) ->
+    LimiterId = {Group, LimiterName},
+    Options = get_limiter_options(LimiterId),
+    ok = set_initial_tokens(Counter, Index, Options),
+    BucketRef = #{counter => Counter, index => Index},
+    ok = emqx_limiter_bucket_registry:insert_bucket(LimiterId, BucketRef),
+    #{
+        counter => Counter,
+        index => Index,
+        correction => 0,
+        last_alloc_time => now_ms_monotonic()
+    }.
+
+alloc_tokens(
```

Comment:
Just a thought. I wonder how would an implementation based on pair of atomics per bucket (i.e. one for tokens, another for "refill epoch", for "self-refilling") compare against this one with timers, e.g.: scalability under high contention and/or massive number of buckets (my guess is the current one should be better up until some large number).